Intersect.Server/Entities/Entity.cs

```diff
@@ -2570,73 +2711,91 @@ public virtual void NotifySwarm(Entity attacker)
 
-        protected byte DirToEnemy(Entity target)
+        protected Direction DirToEnemy(Entity en)
         {
-            //Calculate World Tile of Me
-            var x1 = X + MapController.Get(MapId).MapGridX * Options.MapWidth;
-            var y1 = Y + MapController.Get(MapId).MapGridY * Options.MapHeight;
-
-            //Calculate world tile of target
-            var x2 = target.X + MapController.Get(target.MapId).MapGridX * Options.MapWidth;
-            var y2 = target.Y + MapController.Get(target.MapId).MapGridY * Options.MapHeight;
+            if (en == null)
+            {
+                return Dir;
+            }
 
+            int originY = Y;
+            int originX = X;
+            int targetY = en.Y;
+            int targetX = en.X;
 
-            if (Math.Abs(x1 - x2) > Math.Abs(y1 - y2))
+            // Calculate Y and X offset between origin and target if they're not on the same map instance.
+            if (en.MapInstanceId != MapInstanceId)
             {
-                //Left or Right
-                if (x1 - x2 < 0)
+                if (en.Map.MapGridY < Map.MapGridY)
+                {
+                    originY += Options.MapHeight - 1;
+                }
+                else if (en.Map.MapGridY > Map.MapGridY)
+                {
+                    targetY += Options.MapHeight - 1;
+                }
+
+                if (en.Map.MapGridX < Map.MapGridX)
+                {
+                    originX += Options.MapWidth - 1;
+                }
+                else if (en.Map.MapGridX > Map.MapGridX)
                 {
-                    return (byte) Directions.Right;
+                    targetX += (Options.MapWidth - 1);
                 }
+            }
+
+            // Calculate the offset between origin and target along both of their axis.
+            int yDiff = originY - targetY;
+            int xDiff = originX - targetX;
 
-                return (byte) Directions.Left;
+            // If Y offset is 0, direction is determined by X offset.
+            if (yDiff == 0)
+            {
+                return xDiff > 0 ? Direction.Left : Direction.Right;
             }
 
-            //Left or Right
-            if (y1 - y2 < 0)
+            // If X offset is 0 or If diagonal movement is disabled, direction is determined by Y offset.
+            if (xDiff == 0 || !Options.Instance.MapOpts.EnableDiagonalMovement)
             {
-                return (byte) Directions.Down;
+                return yDiff > 0 ? Direction.Up : Direction.Down;
             }
 
-            return (byte) Directions.Up;
+            // If both X and Y offset are non-zero, direction is determined by both offsets.
+            int xSign = Math.Sign(xDiff);
+            int ySign = Math.Sign(yDiff);
+
+            if (xSign > 0)
+            {
+                return ySign > 0 ? Direction.UpLeft : Direction.DownLeft;
+            }
+
+            return ySign > 0 ? Direction.UpRight : Direction.DownRight;
```

Comment:
See my comment for this code in the client entity code, the sign is unnecessary and the comparisons can be stored ahead of time, also use var